CLDSRV-766: Bucket rate limiting

tmacro opened the pull request:
Note for reviewers: This code has already been individually reviewed in separate PRs. This is just a merge of the feature branch.

tmacro committed:
Add counter and cached config storage

tmacro committed:
impr(CLDSRV-766): Add skeleton

tmacro committed:
impr(CLDSRV-771): Rate limit client wrapper for redis

tmacro committed:
impr(CLDSRV-770): Add unit tests for parseRateLimitConfig

tmacro committed:
impr(CLDSRV-759): Adapt handler code to arsenal changes

tmacro committed:
impr(CLDSRV-759): Add $RATE_LIMIT_SERVICE_USER_ARN to docker-entrypoint.sh

tmacro committed:
impr(CLDSRV-759): Setup rate limit config in CI

tmacro committed:
impr(CLDSRV-759): Add tests for bucketGetRateLimit

tmacro committed:
impr(CLDSRV-760): Add tests for bucketPutRateLimit

tmacro committed:
impr(CLDSRV-761): Add tests for bucketDeleteRateLimit

tmacro committed:
impr(CLDSRV-759): Add rate limit test tooling

anurag4DSB committed:
CLDSRV-755: Add config for Bucket Rate Limiting and tests

anurag4DSB committed:
CLDSRV-755: Add basic (w/o) sync gcra algorithm and tests

anurag4DSB committed:
CLDSRV-762: add check for per bucket rate limit to take node into account

anurag4DSB committed:
CLDSRV-778: Add cleanup job for internal memory ratelimit config

anurag4DSB committed:
CLDSRV-781: Add ratelimit helpers and tests

anurag4DSB committed:
CLDSRV-781: Add ratelimiting pre MD fetch

anurag4DSB committed:
CLDSRV-781: Add ratelimiting post MD fetch

anurag4DSB committed:
CLDSRV-783: Add token reservation system for rate limiting

Implement token reservation architecture where workers request
capacity in advance from Redis. Workers maintain local token
buffers and consume tokens in-memory (no Redis in hot path).
Background refill job requests tokens every 100ms asynchronously.

Redis enforces GCRA atomically at token grant time, ensuring
strict quota enforcement across distributed workers.

Components:
- grantTokens.lua: Redis Lua script for atomic token granting
- tokenBucket.js: WorkerTokenBucket class for local token management
- refillJob.js: Background job for async token replenishment
- client.js: Add grantTokens() method to Redis client

anurag4DSB committed:
CLDSRV-783: Replace GCRA with token consumption in request path

Switch from optimistic GCRA with reconciliation to token
reservation. Workers consume tokens from local buffer instead
of evaluating GCRA per request. This keeps Redis out of the
hot path while enforcing strict quotas.

Changes:
- helpers.js: Use token consumption instead of GCRA evaluation
- server.js: Start token refill job instead of sync job
- cleanup.js: Add token bucket cleanup
- gcra.js: Mark workers parameter unused (kept for compatibility)

anurag4DSB committed:
CLDSRV-783: Remove deprecated/uneeded config fields

Remove syncIntervalMs and reconciliationWeight fields that were
used by the old reconciliation approach. Token reservation uses
a fixed 100ms refill interval instead.

What Each File Changes:

- config.json - Removed syncIntervalMs: 1000
- config.js - Removed parsing for syncIntervalMs and reconciliationWeight
- tokenBucket.js - Linting fixes (line length, return statements, braces)
- cache.js - Updated comments
- tests/config.js - Updated interval calculations for node-level quotas
- tests/gcra.js - Updated tests showing workers are ignored
- tests/helpers.js - Pre-populate token buckets, update assertions

anurag4DSB committed:
CLDSRV-783: Address review comments

anurag4DSB committed:
CLDSRV-762: skip 2 failing tests

anurag4DSB committed:
CLDSRV-783: Rename burstCapacity to bucketSize in GCRA

Use bucketSize parameter name throughout GCRA implementation
and tests for consistency with calling code conventions.

Issue: CLDSRV-783

anurag4DSB committed:
CLDSRV-783: Add unit tests for tokenBucket and refillJob

Add comprehensive unit tests for token reservation components:
- tokenBucket.js: 26 tests covering WorkerTokenBucket class
  and bucket management functions (create, cleanup, stats)
- refillJob.js: 16 tests covering background refill job
  lifecycle (start, stop, error handling)

Total: 42 new tests, bringing rate limit test suite to 137 tests.

Issue: CLDSRV-783

tmacro committed:
impr(CLDSRV-793): Add ensureServiceUser

bert-e commented:
# Hello tmacro,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

BourgoisMickael commented on lib/api/apiUtils/rateLimit/client.js:
```diff
@@ -0,0 +1,78 @@
+
+const { config } = require('../../../Config');
+
+const grantTokensScript = fs.readFileSync(`${__dirname  }/grantTokens.lua`).toString();
```
```suggestion
const grantTokensScript = fs.readFileSync(`${__dirname}/grantTokens.lua`).toString();
```

BourgoisMickael commented on yarn.lock:
```diff
@@ -1451,6 +1451,45 @@ arraybuffer.prototype.slice@^1.0.4:
   optionalDependencies:
     ioctl "^2.0.2"
 
+"arsenal@git+https://github.com/scality/Arsenal#8.2.39":
```
Doesn't match the package.json 8.2.40

BourgoisMickael commented:
Looks like many commits are unsigned

tmacro committed:
CLDSRV-766: Reduce tick log level to trace

tmacro committed:
CLDSRV-794: Fix seeded service user policy

tmacro committed:
CLDSRV-794: Don't check rate limit for rate limit admin routes

tmacro committed:
CLDSRV-794: Skip checking bucket ownership and policies when a service user is making a rate limit admin req

tmacro committed:
CLDSRV-794: Fix lua script

tmacro commented:
/approve

anurag4DSB committed:
CLDSRV-794: Bump arsenal version from 8.2.40 to 8.2.41

This bump brings the change from arsenal where rate limit APIs' action
map are now being exported as part of the arsenal package.

anurag4DSB committed:
CLDSRV-794: use rate limit api action map from arsenal

tmacro committed:
CLDSRV-766: Bump version to 9.2.5

tmacro commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.2`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`
* `development/9.0`
* `development/9.1`



Please check the status of the associated issue CLDSRV-766.

Goodbye tmacro.




*The following options are set:* **approve**
